gateway/assets/script/bootstrap.js

```diff
@@ -306,6 +329,186 @@ app.controller("home", ['$scope', '$log', '$http', '$location', '$interval', '$f
                         $scope.validationError = error1.data;
                     });
             };
+
+            $scope.onEnvInputExpand = function() {
+                $scope.envFieldsVisible = !$scope.envFieldsVisible;
+            }
+
+            $scope.addEnvVar = function(index) {
+                var newInput = {key: "", value: ""};
+                $scope.envVarInputs.splice(index + 1, 0, newInput);
+            }
+
+            $scope.removeEnvVar = function($event, envVar) {
+                var index = $scope.envVarInputs.indexOf(envVar);
+                if ($event.which == 1) {
+                    $scope.envVarInputs.splice(index, 1);
+                }
+            }
+
+            $scope.onSecretInputExpand = function() {
+                $scope.secretFieldsVisible = !$scope.secretFieldsVisible;
+            }
+
+            $scope.addSecret = function(index) {
+                var newInput = {name: ""};
+                $scope.secretInputs.splice(index + 1, 0, newInput);
+            }
+
+            $scope.removeSecret = function($event, secret) {
+                var index = $scope.secretInputs.indexOf(secret);
+                if ($event.which == 1) {
+                    $scope.secretInputs.splice(index, 1);
+                }
+            }
+
+            $scope.onLabelInputExpand = function() {
+                $scope.labelFieldsVisible = !$scope.labelFieldsVisible;
+            }
+
+            $scope.addLabel = function(index) {
+                var newInput = {key: "", value: ""};
+                $scope.labelInputs.splice(index + 1, 0, newInput);
+            }
+
+            $scope.removeLabel = function($event, label) {
+                var index = $scope.labelInputs.indexOf(label);
+                if ($event.which == 1) {
+                    $scope.labelInputs.splice(index, 1);
+                }
+            }
+
+            $scope.onAnnotationInputExpand = function() {
+                $scope.annotationFieldsVisible = !$scope.annotationFieldsVisible;
+            }
+
+            $scope.addAnnotation = function(index) {
+                var newInput = {key: "", value: ""};
+                $scope.annotationInputs.splice(index + 1, 0, newInput);
+            }
+
+            $scope.removeAnnotation = function($event, annotation) {
+                var index = $scope.annotationInputs.indexOf(annotation);
+                if ($event.which == 1) {
+                    $scope.annotationInputs.splice(index, 1);
+                }
+            }
+
+            $scope.envVarInputsToEnvVars = function() {
```

Comment:
I created these conversion functions to allow for displaying the key and value of the objects, then translating them back to objects when they're sent to the server. If anyone knows a way to bind input fields to object keys to remove the need for these additional conversions, I'm all ears!

Reply:
Maybe look into Angular bindings online and see if you can find anything.